src/EFCore.PG/NpgsqlDbFunctionsExtensions.cs

```diff
@@ -159,5 +159,11 @@ var regexPattern
                 RegexOptions.IgnoreCase | RegexOptions.Singleline,
                 _regexTimeout);
         }
+
+        /// <summary>
+        /// Casts the specified <paramref name="target" /> to the PostgreSQL type equivalent
+        /// to <typeparamref name="TTarget" />.
+        /// </summary>
+        public static TTarget Cast<TTarget>(this DbFunctions _, object target) => throw new NotSupportedException();
```

Comment:
It's a useful utility for emitting a cast on the PostgreSQL side that may not have a conversion operator or method defined on the C# side and thus may not be expressible in LINQ directly.

I have a similar method in the EF6 provider, so this is for feature parity. It is not strictly needed for full text search but may be useful for others.

Reply:
I can see the usefulness of this, but it seems like something that belongs in EF Core rather than Npgsql - it has really nothing to do with this specific provider. EFCore.Relational has some "universal" relational translators such as for Like (as well as some others), and it would make sense to have a Cast translator in there as well.

One more note... in this method you're providing the cast target as a CLR type (`TTarget`), which is great. But in some cases I found myself needing to specify a store datatype as a string: when you specify as a CLR type, you're going through the provider's type mappings, which in effect limits you to the default store datatype for that CLR type. For example, there's no way to translate to PostgreSQL type `jsonb`, since its CLR type is string, but the default datatype for that is actually `text`.

I recently had this issue when fixing #319, where I needed to cast to `citext`, and added my own `ExplicitStoreTypeCastExpression`, which is the same as an `ExplicitCastExpression` but which is provided a store datatype as a string instead of a CLR type, effectively bypassing the type mappings. This also probably belongs in EF Core rather than the provider.

If you end up doing a PR to EF Core, you may want to add the string-based cast option as well.

Reply:
@roji Seems related to https://github.com/aspnet/EntityFrameworkCore/issues/4978 which started off being about Unicode, but in general is about providing type mappings for parts of a query where it is not possible/easy to infer the mapping from some related property. /cc @divega

Reply:
Thanks @ajcvickers. I do think it also makes sense to provide some means for users to force a cast sometimes, as a last resort or workaround. As I said above, I ended up needing a manual explicit cast to a store type (not a CLR type - to bypass the type mapping system), you can [check this code out](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/5488caf054f3b56c2b8ab33b34221d6aeffa8619/src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlStringStartsWithTranslator.cs#L73).

Reply:
@roji @rwasef1830 @ajcvickers still don't understand how `EF.Functions.Cast<X>(y)` would be different form `(X)y`. Is it different because the latter may be removed or ignored by EF Core?

Reply:
```(X)y``` will not compile if there is no C# implementation of a cast operator or a built-in conversion. The method call will be allowed to compile. There is no difference in case there *is* a conversion, but that's not the point of the method.

Reply:
@divega See https://github.com/aspnet/EntityFrameworkCore/issues/11089 where the answer was to force an "invalid" cast into the model using `(string)(object)myInt`. This works because it is never executed, but the cast call seems cleaner. But my main point was that something like this could allow facets to be changed.

Reply:
Keep in mind that there are also store types which simply don't exist in C#. For example, CLR string maps to both text and json in PostgreSQL, with text being the default. But in some scenarios I may need to specify a cast to JSON specifically (since PostgreSQL doesn't do implicit casting in several scenarios).

Reply:
Ok, thanks, I think get it. `EF.Functions.Cast<X>(y)` is useful if the three conditions below are true:
1. There is no valid conversion
2. You don't want to write `(X)(object)y`
3. The X CLR type provides enough information (e.g. it is not about explicitly casting to store type or specifying a facet

I think it is interesting, and we could have something like this in EF Core, but perhaps it is more interesting to consider other overloads or variations that can set various facets or the store type. I think it is unlikely that adding this now would prevent us from adding the other variations we want later.

Reply:
@rwasef1830, I guess you want to open an issue for this feature in the EF Core repo, and remove this commit from the PR?

Reply:
OK, got caught up and busy at the moment, will get it done next week hopefully.